### PR TITLE
Add zip-plugin-builds command

### PR DESCRIPTION
**Pull request by westonruter**
_This is a sub-PR of #1227. Merge it before merging this PR._

In order to facilitate testing of pre-release builds, it is very helpful to have ZIP archives of the plugins to manually install on a WordPress site. This PR adds a `npm run zip-plugin-builds` command which is run automatically as part of `npm run build-plugins`. The ZIP files are located in the `build` directory alongside the built plugin directories.

### CODE REVIEW

**Comment by thelovekesh** [comment]
Should this process be a part of webpack?

<details><summary>[example diff]</summary>

```diff
diff --git a/bin/webpack/utils.js b/bin/webpack/utils.js
index d0a909df..629dbdfa 100644
--- a/bin/webpack/utils.js
+++ b/bin/webpack/utils.js
@@ -1,5 +1,6 @@
 const fs = require( 'fs' );
 const path = require( 'path' );
+const { execSync } = require( 'child_process' );
 
 /**
  * Return plugin root path.
@@ -97,10 +98,24 @@ const assetDataTransformer = ( content, absoluteFrom ) => {
 	return `<?php return array('dependencies' => array(), 'version' => '${ version }');`;
 };
 
+/**
+ * Create plugins zip file using `zip` command.
+ *
+ * @param {string} pluginPath The path where the plugin build is located.
+ * @param {string} pluginName The name of the plugin.
+ *
+ * @return {void}
+ */
+const createPluginZip = ( pluginPath, pluginName ) => {
+	const command = `cd ${ pluginPath } && zip -r ${ pluginName }.zip ${ pluginName }`;
+	execSync( command );
+};
+
 module.exports = {
 	getPluginRootPath,
 	deleteFileOrDirectory,
 	getPluginVersion,
 	generateBuildManifest,
 	assetDataTransformer,
+	createPluginZip,
 };
diff --git a/package.json b/package.json
index c5c9405c..92e8c338 100644
--- a/package.json
+++ b/package.json
@@ -30,15 +30,15 @@
     "readme": "./bin/plugin/cli.js readme",
     "versions": "./bin/plugin/cli.js versions",
     "build": "wp-scripts build",
-    "build-plugins": "npm-run-all 'build:plugin:!(performance-lab)' 'zip-plugin-builds'",
-    "build:plugin:performance-lab": "rm -rf build/performance-lab && mkdir -p build/performance-lab && git archive HEAD | tar -x -C build/performance-lab",
+    "build-plugins": "npm-run-all 'build:plugin:!(performance-lab)'",
+    "build:plugin:performance-lab": "rm -rf build/performance-lab && mkdir -p build/performance-lab && git archive HEAD | tar -x -C build/performance-lab && cd build && zip -qr performance-lab.zip performance-lab",
     "build:plugin:auto-sizes": "webpack --mode production --env plugin=auto-sizes",
     "build:plugin:dominant-color-images": "webpack --mode production --env plugin=dominant-color-images",
     "build:plugin:embed-optimizer": "webpack --mode production --env plugin=embed-optimizer",
     "build:plugin:optimization-detective": "webpack --mode production --env plugin=optimization-detective",
     "build:plugin:speculation-rules": "webpack --mode production --env plugin=speculation-rules",
     "build:plugin:webp-uploads": "webpack --mode production --env plugin=webp-uploads",
-    "zip-plugin-builds": "bin/zip-plugin-builds.sh",
+    "zip-plugin-builds": "npm-run-all 'build:plugin:!(performance-lab) --env zip=true' 'build:plugin:performance-lab'",
     "generate-pending-release-diffs": "bin/generate-pending-release-diffs.sh",
     "format-js": "wp-scripts format",
     "lint-js": "wp-scripts lint-js",
diff --git a/webpack.config.js b/webpack.config.js
index 2de1217b..a2c6cc79 100644
--- a/webpack.config.js
+++ b/webpack.config.js
@@ -10,6 +10,7 @@ const CopyWebpackPlugin = require( 'copy-webpack-plugin' );
  */
 const { plugins: standalonePlugins } = require( './plugins.json' );
 const {
+	createPluginZip,
 	assetDataTransformer,
 	deleteFileOrDirectory,
 	generateBuildManifest,
@@ -102,7 +103,8 @@ const buildPlugin = ( env ) => {
 		};
 	}
 
-	const to = path.resolve( __dirname, 'build', env.plugin );
+	const buildDir = path.resolve( __dirname, 'build' );
+	const to = path.resolve( buildDir, env.plugin );
 	const from = path.resolve( __dirname, 'plugins', env.plugin );
 	const dependencies = pluginsWithBuild.includes( env.plugin )
 		? [ `${ env.plugin }` ]
@@ -138,6 +140,11 @@ const buildPlugin = ( env ) => {
 					// After emit, generate build manifest.
 					compiler.hooks.afterEmit.tap( 'AfterEmitPlugin', () => {
 						generateBuildManifest( env.plugin, from );
+
+						// If zip flag is passed, create a zip file.
+						if ( env.zip ) {
+							createPluginZip( buildDir, env.plugin );
+						}
 					} );
 				},
 			},

```

</details>

**Comment by westonruter** [comment]
@thelovekesh I guess so. It's just I know Bash, not Webpack.

**Comment by thelovekesh** [comment]
I suggest incorporating this so that building the zip file becomes part of the asset-building lifecycle itself. And the above diff should work as expected.

**Comment by westonruter** [comment]
Feel free to amend my PR with a revert of my Bash script and inclusion your patch

File: .github/CODEOWNERS
Comment (thelovekesh) [inline .github/CODEOWNERS:3]: I added @thelovekesh to `bin` codeowners.

File: .github/CODEOWNERS
Comment (thelovekesh) [inline .github/CODEOWNERS:11]: Fixed paths and include server timing API files.

File: webpack.config.js
Comment (thelovekesh) [inline webpack.config.js:29]: During each build step, Webpack creates an empty `dist` directory when `output: {}` is specified. This update prevents that and also addresses code duplication in several places below.

File: package.json
Comment (thelovekesh) [inline package.json]: TODO: Remove it once https://github.com/WordPress/performance/pull/1182 is merged.

**Comment by thelovekesh** [comment]
@westonruter Can you please test it once?

File: .github/CODEOWNERS
Comment (westonruter) [reply .github/CODEOWNERS:3]: Who?

**Comment by westonruter** [comment]
@thelovekesh Looks good! Tested and it works well.

**Comment by westonruter** [comment]
@thelovekesh I guess you'll have to approve this pull request since I'm the one who opened it originally. But before we merge, we'll need to merge #1227 first.

**Review by thelovekesh — APPROVED** [review]

**Comment by westonruter** [comment]
@thelovekesh I got the PHPStan error when attempting to commit the merge conflict:

```
Script phpstan analyse --memory-limit=2048M handling the phpstan event returned with error code 1
 ------ ---------------------------------------------------------------------- 
  Line   tests/plugins/performance-lab/load-tests.php                          
 ------ ---------------------------------------------------------------------- 
  35     Call to method PHPUnit\Framework\Assert::assertFalse() with int will  
         always evaluate to false.                                             
  50     Call to method PHPUnit\Framework\Assert::assertFalse() with int will  
         always evaluate to false.                                             
  73     Call to method PHPUnit\Framework\Assert::assertFalse() with int will  
         always evaluate to false.                                             
  144    Call to method PHPUnit\Framework\Assert::assertFalse() with int will  
         always evaluate to false.                                             
 ------ ---------------------------------------------------------------------- 
```

It may be due to the usage with `lint-staged` in which it is passing individual files to check rather than the entire repo, because once I forced the commit and ran `npm run phpstan`, then there was no error. 

I can reproduce the issue with the same command that `lint-staged` invoked:

```bash
npm run phpstan -- performance.php plugins/dominant-color-images/class-dominant-color-image-editor-imagick.php plugins/optimization-detective/optimization.php plugins/performance-lab/includes/admin/load.php plugins/performance-lab/includes/admin/plugins.php plugins/performance-lab/includes/admin/server-timing.php plugins/performance-lab/includes/server-timing/class-perflab-server-timing-metric.php plugins/performance-lab/includes/server-timing/class-perflab-server-timing.php plugins/performance-lab/includes/server-timing/defaults.php plugins/performance-lab/includes/server-timing/hooks.php plugins/performance-lab/includes/server-timing/load.php plugins/performance-lab/includes/server-timing/object-cache.copy.php plugins/performance-lab/includes/site-health/audit-autoloaded-options/helper.php plugins/performance-lab/includes/site-health/audit-autoloaded-options/hooks.php plugins/performance-lab/includes/site-health/audit-enqueued-assets/helper.php plugins/performance-lab/includes/site-health/audit-enqueued-assets/hooks.php plugins/performance-lab/includes/site-health/avif-support/helper.php plugins/performance-lab/includes/site-health/avif-support/hooks.php plugins/performance-lab/includes/site-health/load.php plugins/performance-lab/includes/site-health/webp-support/helper.php plugins/performance-lab/includes/site-health/webp-support/hooks.php plugins/performance-lab/load.php plugins/speculation-rules/helper.php plugins/speculation-rules/load.php plugins/webp-uploads/deprecated.php plugins/webp-uploads/settings.php tests/bootstrap.php tests/plugins/auto-sizes/auto-sizes-test.php tests/plugins/performance-lab/includes/admin/load-tests.php tests/plugins/performance-lab/includes/admin/server-timing-tests.php tests/plugins/performance-lab/includes/server-timing/load-tests.php tests/plugins/performance-lab/includes/server-timing/perflab-server-timing-metric-tests.php tests/plugins/performance-lab/includes/server-timing/perflab-server-timing-tests.php tests/plugins/performance-lab/includes/site-health/audit-autoloaded-options/audit-autoloaded-options-test.php tests/plugins/performance-lab/includes/site-health/audit-enqueued-assets/audit-enqueued-assets-helper-test.php tests/plugins/performance-lab/includes/site-health/audit-enqueued-assets/audit-enqueued-assets-test.php tests/plugins/performance-lab/includes/site-health/audit-enqueued-assets/class-audit-assets-transients-set.php tests/plugins/performance-lab/includes/site-health/audit-enqueued-assets/class-site-health-mock-responses.php tests/plugins/performance-lab/load-tests.php tests/plugins/webp-uploads/helper-tests.php tests/plugins/webp-uploads/image-edit-tests.php tests/plugins/webp-uploads/load-tests.php tests/plugins/webp-uploads/rest-api-tests.php tests/utils/TestCase/DominantColorTestCase.php
```

The issue I think is that we shouldn't be running `lint-staged` on individual paths. We avoided doing this for the AMP plugin in the `.lint-staged.js` as [follows](https://github.com/ampproject/amp-wp/blob/235e83e97518ddc853ed7028972a5b7e072fa0be/.lintstagedrc.js#L7):

```js
'*.php': () => 'composer analyze',
```

I think we can do the same once we finally pick up the lint-staged work again in https://github.com/WordPress/performance/pull/1090.

**Comment by thelovekesh** [comment]
Makes sense. Additionally, PHPStan recommends running the analysis on the entire codebase rather than on specific paths. This is because changes in one file might affect other parts of the project, potentially hiding errors if only a subset of files is analyzed.

**Pull request merged**
